pkg/spoke/controllers/deletioncontroller/deletion_controller.go

```diff
@@ -112,41 +134,95 @@ func (m *StaleManifestDeletionController) syncManifestWork(ctx context.Context,
-	// delete applied resources which are no longer maintained by manifest work
-	noLongerMaintainedResources := findUntrackedResources(manifestWork.Status.AppliedResources, appliedResources)
-	if errs := m.deleteAppliedResources(noLongerMaintainedResources); len(errs) != 0 {
-		return utilerrors.NewAggregate(errs)
+	// reset the rate limiter for the manifest work
+	if len(deletingResources) == 0 {
+		forgetWork()
 	}
 
-	// update work status with latest applied resources. if this conflicts, we'll simply try again later
+	// update work status with latest applied resources. if this conflicts, we'll try again later
 	// for retrying update without reassessing the status can cause overwriting of valid information.
 	manifestWork.Status.AppliedResources = appliedResources
 	_, err := m.manifestWorkClient.UpdateStatus(ctx, manifestWork, metav1.UpdateOptions{})
 	return err
 }
 
-// deleteAppliedResources deletes all applied resources in the given resource meta array.
-func (m *StaleManifestDeletionController) deleteAppliedResources(resources []workapiv1.AppliedManifestResourceMeta) []error {
+// deleteAppliedResources deletes all given applied resources and returns those
+// which are deleted but not finalized yet
+func (m *StaleManifestDeletionController) deleteAppliedResources(resources []workapiv1.AppliedManifestResourceMeta) ([]workapiv1.AppliedManifestResourceMeta, []error) {
+	var deletingResources []workapiv1.AppliedManifestResourceMeta
 	var errs []error
+
 	for _, resource := range resources {
 		gvr := schema.GroupVersionResource{Group: resource.Group, Version: resource.Version, Resource: resource.Resource}
-		err := m.spokeDynamicClient.Resource(gvr).Namespace(resource.Namespace).Delete(context.TODO(), resource.Name, metav1.DeleteOptions{})
-		switch {
-		case errors.IsNotFound(err):
-			// no-oop
-		case err != nil:
-			errs = append(errs, fmt.Errorf("Failed to delete resource %v with key %s/%s: %w", gvr, resource.Namespace, resource.Name, err))
+		u, err := m.spokeDynamicClient.
+			Resource(gvr).
+			Namespace(resource.Namespace).
+			Get(context.TODO(), resource.Name, metav1.GetOptions{})
+		if errors.IsNotFound(err) {
+			klog.V(2).Infof("Resource %v with key %s/%s is removed Successfully", gvr, resource.Namespace, resource.Name)
 			continue
 		}
-		klog.V(2).Infof("Successfully delete resource %v with key %s/%s", gvr, resource.Namespace, resource.Name)
+
+		if err != nil {
+			errs = append(errs, fmt.Errorf(
+				"Failed to get resource %v with key %s/%s: %w",
+				gvr, resource.Namespace, resource.Name, err))
+			deletingResources = append(deletingResources, resource)
+			continue
+		}
+
+		// leave deleting resource untouched
+		if deletionTimestamp := u.GetDeletionTimestamp(); deletionTimestamp != nil && !deletionTimestamp.IsZero() {
+			// set the UID even if the resource is deleted by other client
+			if resource.UID == "" {
+				resource.UID = string(u.GetUID())
+			}
+			deletingResources = append(deletingResources, resource)
```

Comment:
check to see if you're deleting the UID you expect.  Otherwise you could be waiting for a different instance to be deleted.  if UIDs don't match, then you're done


```
pendingFinalization := u.GetDeletionTimestamp() != nil && !u.GetDeletionTimestamp().IsZero()
if pendingFinalization {
  waitingForFinalization := len(resource.UID) > 0
  if waitingForFinalization && resource.UID != u.GetUID{
    // the instance is deleted, so do not add to the resourcesPendingDeletion list
    continue
  }

  // if we aren't waiting for finalization or the UID does match, then we want to ensure we add it to the pending list
  copyResource
  copyResource.UID = u.GetUID
  deletingResources = append(deletingResources, resource)
  continue
}